File: lua-project/src/logic/CannonBall01AI.lua

local creator = cc.import("creator.init")

local CannonBall01AI = cc.class("CannonBall01AI", creator.ComponentBase)

local setpos = cc.Node.setPosition

function CannonBall01AI:start(target)
    self.hittime = target.hittime
    self.hitx    = target.hitx
    self.hity    = target.hity

    local x, y = target:getPosition()
    self.x    = x
    self.y    = y
    self.time = 0
    self.g    = -1000
    self.over = false

    local ht = self.hittime
    self.ox = (self.hitx - self.x) / ht
    self.oy = ((self.hity - self.y) - ((self.g * ht) * (ht / 2))) / ht
end

function CannonBall01AI:update(target, dt)
    if self.over then return end

    self.time = self.time + dt
    local time = self.time

    local x = self.x + time * self.ox
    local y = self.y + time * self.oy + self.g * time * time / 2
    setpos(target, x, y)

    if time >= self.hittime then
        self.over = true
        self:boom(target, x, y)
    end
end

function CannonBall01AI:boom(target, x, y)
    local assets = creator.getAssets()
    local boom = assets:createPrefab("resources/CannonBoom0001")
    boom:setPosition(x, y)
    boom:setLocalZOrder(2000)
    boom:trackComponents()

    local animation = boom:getComponent("cc.Animation")
    animation:play(boom, function()
        boom:cleanupComponets()
        boom:removeFromParent()
    end)
    target:getParent():addChild(boom)

    target:cleanupComponets()
    target:removeFromParent()
end

return CannonBall01AI

File: creator-project/assets/Script/CannonBall01AI.js
cc.Class({
    extends: cc.Component,

    properties: {
        // foo: {
        //    default: null,
        //    url: cc.Texture2D,  // optional, default is typeof default
        //    serializable: true, // optional, default is true
        //    visible: true,      // optional, default is true
        //    displayName: 'Foo', // optional
        //    readonly: false,    // optional, default is false
        // },
        // ...
        boomPrefab: {
            default: null,
            type: cc.Prefab
        },
        hitx: 0,
        hity: 0,
        hittime: 0,
        g: -1000
    },

    // use this for initialization
    onLoad: function () {
        var pos = this.node.getPosition();
        this.x = pos.x;
        this.y = pos.y;
        this.time = 0;
        this.over = false;
        
        var ht = this.hittime;
        this.ox = (this.hitx - this.x) / ht;
        this.oy = ((this.hity - this.y) - ((this.g * ht) * (ht / 2))) / ht;
    },

    // called every frame, uncomment this function to activate update callback
    update: function (dt) {
        if (this.over) return;

        this.time = this.time + dt;
        var time = this.time;
    
        var x = this.x + time * this.ox;
        var y = this.y + time * this.oy + this.g * time * time / 2;
        var pos = cc.p(x, y);
        this.node.setPosition(pos);

        if (time >= this.hittime) {
            this.over = true
            this.boom(pos);
        }
    },
    
    boom: function(pos) {
        var boom = cc.instantiate(this.boomPrefab);
        boom.setPosition(pos);
        boom.setLocalZOrder(2000);
        this.node.getParent().addChild(boom);
        boom.getComponent(cc.Animation).play();

        this.node.destroy();
    }
});
